internal/integrationtest/util.go

```diff
@@ -0,0 +1,257 @@
+// apply to cover connection reuse.
+func (s TransportSpec) TestReuseConnRoundRobin(t *testing.T) {
+	var wg sync.WaitGroup
+	count := 10
+
+	server, addr := s.NewServer(t, ":0")
+	defer server.Stop()
+
+	client, c := s.NewClient(t, []string{addr})
+	defer client.Stop()
+
+	wg.Add(count)
+	call := func() {
+		defer wg.Done()
+		ctx := context.Background()
+		ctx, cancel := context.WithTimeout(ctx, 50*time.Millisecond)
+		defer cancel()
+		assert.NoError(t, Call(ctx, c))
+	}
+	for i := 0; i < count; i++ {
+		go call()
+		time.Sleep(10 * time.Millisecond)
+	}
+
+	wg.Wait()
+}
+
+// TestBackoffConnRoundRobin is a reusable test that any transport can apply to
+// cover connection management backoff.
+func (s TransportSpec) TestBackoffConnRoundRobin(t *testing.T) {
+	var wg sync.WaitGroup
+	addr := "127.0.0.1:31782"
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		client, c := s.NewClient(t, []string{addr})
+		defer client.Stop()
+
+		ctx := context.Background()
+		ctx, cancel := context.WithTimeout(ctx, time.Second)
+		defer cancel()
+
+		// Eventually succeeeds, when the server comes online.
+		assert.NoError(t, Call(ctx, c))
+	}()
+
+	// Give the client time to make multiple connection attempts.
+	time.Sleep(10 * time.Millisecond)
+	server, _ := s.NewServer(t, addr)
+	defer server.Stop()
+
+	wg.Wait()
+}
+
+// TestReconnRoundRobin is a reusable test that exercises any
+// transport's ability to reconnect to a peer if it is temporarily unavailable
+// while being retained.
+func (s TransportSpec) TestReconnRoundRobin(t *testing.T) {
+	server, addr := s.NewServer(t, ":0")
+	// server.Stop() is explicit in this test.
+
+	client, c := s.NewClient(t, []string{addr})
+	defer client.Stop()
+
+	// Induce a connection
+	func() {
+		ctx := context.Background()
+		ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+		defer cancel()
+		assert.NoError(t, Call(ctx, c))
+	}()
+
+	// Stop the server so a subsequent request must fail
+	server.Stop()
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		CallUntilSuccess(t, c, 100*time.Millisecond)
+	}()
+
+	// Restart the server so it can reconnect.
+	time.Sleep(20 * time.Millisecond)
+	restoredServer, _ := s.NewServer(t, addr)
+	defer restoredServer.Stop()
+
+	wg.Wait()
+}
+
+// Blast sends a blast of calls to the client and verifies that they do not
+// err.
+func Blast(ctx context.Context, t *testing.T, c raw.Client) {
+	for i := 0; i < 10; i++ {
+		assert.NoError(t, Call(ctx, c))
+	}
+}
+
+// CallUntilSuccess sends a request until it succeeds.
+func CallUntilSuccess(t *testing.T, c raw.Client, interval time.Duration) {
+	for {
+		ctx := context.Background()
+		ctx, cancel := context.WithTimeout(ctx, interval)
+		if err := Call(ctx, c); err == nil {
+			cancel()
+			break
+		}
+		cancel()
+	}
+}
+
+// Call sends an echo request to the client.
+func Call(ctx context.Context, c raw.Client) error {
+	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
```

Comment:
Today I learned: WithTimeout does not override the timeout, it takes the min of the passed in timeout and the current context timeout.